eos-tech-support/eos-fix-ostree-repo

```diff
@@ -52,6 +52,77 @@ import sys
 import time
 
 
+# Prior to ostree-2017.1, the GI annotation for ostree_repo_list_objects
+# was wrong, making it unusable from bindings[1]. This is tricky to
+# detect since the version checking was not added until ostree-2017.3.
+# However, in the same commit,
+# ostree_repo_list_commit_objects_starting_with was fixed so that the
+# out_commits parameter was marked properly.
+#
+# Use the direction of this argument as a proxy to know when
+# ostree_repo_list_objects will work. Otherwise, make our own
+# list_objects implementation and monkey-patch it into the Repo class.
+#
+# 1. https://github.com/ostreedev/ostree/commit/300752e5
+_func = OSTree.Repo.list_commit_objects_starting_with
+if _func.get_arguments()[1].get_direction() == gi._gi.Direction.IN:
+    import glob
+
+    # Fake ostree_repo_list_objects. See
+    # https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-repo.c
+    # for the real implementation.
+    def _list_objects(self, flags, cancellable=None):
+        objects = {}
+        repo_path = self.get_path().get_path()
+        repo_mode = self.get_mode()
+
+        # Objects live in the objects directory split after the 2nd
+        # character of their sha256sum. E.g.,
+        # 8d/2925839245dd91ac9fbfcc0e7a383cddf5145bd7c5bc5de0d46929a3fa5963.file.
+        objdir_pattern = repo_path + '/objects/[a-f0-9][a-f0-9]'
+        for objdir in glob.iglob(objdir_pattern):
+            for entry in os.listdir(objdir):
+                name, ext = os.path.splitext(entry)
+
+                if len(name) != 62:
+                    # Not a partial sha256
+                    continue
+
+                if ext == '':
+                    continue
+                elif (ext == '.filez' and
+                      repo_mode == OSTree.RepoMode.ARCHIVE_Z2):
+                    objtype = OSTree.ObjectType.FILE
+                elif (ext == '.file' and
+                      repo_mode != OSTree.RepoMode.ARCHIVE_Z2):
```

Comment:
Technically that should be `and repo_mode in (OSTree.RepoMode.BARE, .BARE_USER, .BARE_USER_ONLY)`. The expressions are equivalent for the moment, but they might change over time.

Then again, some of those `RepoMode`s probably didn’t exist in the version of OSTree in EOS 3.0. Sigh. Probably best to leave it as you’ve written it then.

Reply:
Yeah. I could make a copy of ostree's internal `mode_is_bare` helper and only match on whatever's in the `RepoMode` enum besides `ARCHIVE_Z2`. Hmm, that looks kind of ugly as I think you have to look at the internal `__enum_values__` property.

Reply:
Exactly. I think it’s probably best left as you’ve written it.